### PR TITLE
gh-116720: Fix corner cases of taskgroups

### PR DESCRIPTION
**Pull request by gvanrossum**
- [x] Make test_timeouts pass (test_timeout_taskgroup)
- [x] Add tests derives from examples in the issue
- [x] Add test for new `uncancel()` behavior (reset `_must_cancel`)
- [x] Add news entry
- [x] Add What's New entry
- [x] Update docs for `uncancel()`
- [x] Update docs for task groups
- [ ] Add Co-Authored-By: @arthur-tacca

* Issue: gh-116720

### CODE REVIEW

**Comment by gvanrossum** [comment]
@arthur-tacca, what do you think of this PR? Does it pass all your tests? Do you want to contribute in any way (e.g. by converting your example programs into unittests)?

**Comment by gvanrossum** [comment]
@Tinche Same question for you -- you might have an opinion on the changes that I'm proposing to make to asyncio task and task group behavior here.

**Comment by agronholm** [comment]
> @agronholm I wonder if you could review this PR? (It's in draft mode because there are no tests or doc yet, but the changes to taskgroups.py and tasks.py are as I plan to keep them.)
> 
> It might affect anyio because of two things: (a) Better support for nested taskgroups; (b) change to uncancel() to **reset** the internal "must cancel" flag when the pending cancellation count reaches zero.

Sure, I can take a look.

**Comment by agronholm** [comment]
It's hard to definitively say if this will affect AnyIO, but I did run the task group tests against this PR and they passed. The only place where AnyIO looks at this flag is in `CancelScope._deliver_cancellation()` where it will skip explicitly cancelling a task if this flag is already set.

**Comment by gvanrossum** [comment]
Sorry, which flag is that? ‘cancelling()’ or ‘_must_cancel’?

File: Lib/asyncio/tasks.py [hunk 1]
```diff
@@ -255,6 +255,8 @@ def uncancel(self):
         """
         if self._num_cancels_requested > 0:
             self._num_cancels_requested -= 1
+            if self._num_cancels_requested == 0:
+                self._must_cancel = False
```
Comment (agronholm) [inline Lib/asyncio/tasks.py:259]: Shouldn't this also clear the cancel message?

**Comment by agronholm** [comment]
> Sorry, which flag is that? ‘cancelling()’ or ‘_must_cancel’?

`_must_cancel`.

File: Lib/asyncio/tasks.py (hunk 1 above)
Comment (gvanrossum) [reply Lib/asyncio/tasks.py:259]: I doubt it -- `self._cancel_message` is never cleared, and it's never used unless `self._must_cancel` is set.

Comment (agronholm) [reply Lib/asyncio/tasks.py:259]: Okay – I was just confused by that `Py_CLEAR()` in the C version, but this explains it.

Comment (gvanrossum) [reply Lib/asyncio/tasks.py:259]: Oh, you're right, that CLEAR is unnecessary there too. :-)

**Comment by gvanrossum** [comment]
Still to do are docs updates for uncancel and for task groups. Also, I'd love help with shortening the news blurb.

**Comment by gvanrossum** [comment]
@willingc Would you be my reviewer for this PR? I feel particularly challenged by the various documentation updates -- this is a pretty subtle improvement (given how long it's taken before someone reported the issue :-) and I'm struggling describing the changes and the end state appropriately in various places:

- Changelog (Misc/NEWS)
- What's new in 3.13
- asyncio-task.rst: uncancel docs and task group docs

I'd say that currently the Changelog news item is probably too long, and there's too much duplication between all three forms of documentation.

@arthur-tacca: I haven't heard from you on the issue in a while. Your feedback on any part of this PR would be much appreciated. I plan to give you credit by adding `Co-Authored-By: Arthur Tacca` to the final commit message, since I cribbed most of the unit tests from your original demonstrations. I will also add ``(Inspired by an issue reported by Arthur Tacca in :gh:`116720`.)`` to the "what's new" section.

File: Lib/asyncio/taskgroups.py [hunk 2]
```diff
@@ -139,6 +140,11 @@ async def __aexit__(self, et, exc, tb):
             self._errors.append(exc)
 
         if self._errors:
+            # If the parent task is being cancelled from the outside,
```
Comment (willingc) [inline Lib/asyncio/taskgroups.py]:
```suggestion
            # If the parent task is being cancelled from the outside of the taskgroup,
```

File: Lib/asyncio/taskgroups.py [hunk 3]
```diff
@@ -139,6 +140,11 @@ async def __aexit__(self, et, exc, tb):
...
+            # re-cancel it, while keeping the cancel count stable.
```
Comment (willingc) [inline Lib/asyncio/taskgroups.py]:
```suggestion
            # un-cancel and re-cancel the parent task, which will keep the cancel count stable.
```

File: Doc/library/asyncio-task.rst [hunk 4]
```diff
@@ -1369,6 +1391,15 @@ Task Object
       catching :exc:`CancelledError`, it needs to call this method to remove
       the cancellation state.
 
+      When this method decrements the cancellation count to zero,
+      if a previous :meth:`cancel` call had arranged for a
```
Comment (willingc) [inline Doc/library/asyncio-task.rst]:
```suggestion
      the method checks if a previous :meth:`cancel` call had arranged for
```

File: Doc/library/asyncio-task.rst [hunk 5]
```diff
@@ -1369,6 +1391,15 @@ Task Object
...
+      :exc:`CancelledError` to be thrown into the task,
```
Comment (willingc) [inline Doc/library/asyncio-task.rst]:
```suggestion
      :exc:`CancelledError` to be thrown into the task.
```

File: Doc/library/asyncio-task.rst [hunk 6]
```diff
@@ -1369,6 +1391,15 @@ Task Object
...
+      but this hadn't been done yet, that arrangement will be
```
Comment (willingc) [inline Doc/library/asyncio-task.rst]:
```suggestion
      If this hadn't been done yet, that arrangement will be
```

File: Doc/whatsnew/3.13.rst [hunk 7]
```diff
@@ -296,6 +289,34 @@ asyncio
...
   (Contributed by Justin Arthur in :gh:`77714`.)
 
+* When :func:`asyncio.TaskGroup.create_task` is called on an inactive
+  :class:`asyncio.TaskGroup`, the given coroutine will be closed (which
+  prevents a :exc:`RuntimeWarning` about the given coroutine being
+  never awaited).
+  (Contributed by Arthur Tacca and Jason Zhang in :gh:`115957`.)
+
+* Improved behavior of :class:`asyncio.TaskGroup` when an outside cancellation
+  collides with an internal cancellation. For example, when two task groups
```
Comment (willingc) [inline Doc/whatsnew/3.13.rst:299]: Nice!

**Review by willingc — APPROVED** [review]
@gvanrossum Overall, this looks good. I made a few suggestions on wording, but the existing text is also fine. Thanks!

**Review by willingc — APPROVED** [review]
Looks good. Thanks everyone.

**Comment by Tinche** [comment]
@gvanrossum would you mind if I copied your solution into quattro to serve as a backport of sorts, for 3.11 and 3.12? Quattro is Apache 2 licensed, that's compatible right?

**Comment by gvanrossum** [comment]
> @gvanrossum would you mind if I copied your solution into quattro to serve as a backport of sorts, for 3.11 and 3.12? Quattro is Apache 2 licensed, that's compatible right?

That should be absolutely fine -- if I understand the PSF license correctly, all you need to do is add this text somewhere
```
Copyright © 2001-2024 Python Software Foundation; All Rights Reserved
```
and add a copy of the PSF license for 3.13 somewhere.

**Pull request merged**